exec/cnex/number.c

```diff
@@ -17,49 +17,88 @@
+    memset(buf, 0x00, len);
+    const int PRECISION = 34;
 
     assert(len != 0);
 
-    char val[50];
-    char lead[50];
+    char val[50] = { 0 };
     bid128_to_string(val, x);
 
-    char *s = &val[0];
-
-    /* Skip over the leading +, if there is one. */
-    while(*s != '\0' && (*s == '+')) s++;
-
-    char *v = s;
+    char *v, *s = v = &val[1];
 
-    while (*s != '\0' && (*s != 'E')) s++;
+    // Calculate the length of the string, locating the exponent value at the same time.
+    int slen = 0;
+    while (*s != '\0' && (*s != 'E')) s++, slen++;
 
     char *E = s;
 
-    if (*s == 'E') {
-        /* Deal with the exponent. */
-        int i;
-        iLeadingZeros = abs(atoi(++s));
-        for (i = 0; i < iLeadingZeros; i++) {
-            lead[i] = '0';
-        }
-        lead[i++] = '.';
-        lead[i++] = '\0';
-        strcpy(buf, lead);
+    if (*s != 'E') {
+        buf = val;
+        return;
     }
 
+    int exponent = atoi(++s);
     *E = '\0';
-    strcat(buf, v);
+    char *p = E - 1;
+
+    while ((p > v) && (*p == '0')) p--;
+
+    const int last_significant_digit = (int)(p - v) + 1;
+    const int trailing_zeros = (int)slen - last_significant_digit;
+    exponent += trailing_zeros;
+    char sbuf[50] = { 0 };
+    char r[50] = { 0 };
+    memcpy(sbuf, v, last_significant_digit);
+    slen = last_significant_digit;
+    sbuf[last_significant_digit+1] = '\0';
+
+    if (exponent != 0) {
+        if (exponent > 0 && slen + exponent <= PRECISION+1) {
+            memcpy(r, sbuf, slen);
+            for (int i = slen; i < (slen + exponent); i++) {
+                r[i] = '0';
+            }
+        } else if (exponent < 0 && -exponent < slen) {
+            memcpy(r, &sbuf[0], slen+exponent);
+            r[slen+exponent] = '.';
+            memcpy(&r[slen+exponent]+1, &sbuf[slen+exponent], slen - (slen+exponent));
+        } else if (exponent < 0 && -exponent == slen-1) {
+            memcpy(r, "0.", 2);
+            memcpy(&r[2], sbuf, slen);
+        } else if (exponent < 0 && slen - exponent <= PRECISION+2) {
+            int count = 2;
+            memcpy(r, "0.", count);
+            for (int i = 0; i < -exponent-slen;i++, count++) {
+                r[count] = '0';
+            }
+            memcpy(&r[count], sbuf, slen);
+            r[count+slen] = '\0';
+        } else {
+            exponent += slen - 1;
+            if (slen >= 3) {
+                memcpy(r, sbuf, 1);
+                r[1] = '.';
+                memcpy(&r[2], &sbuf[1], last_significant_digit-1);
+            }
+            char exp[36];
+            int explen = sprintf(exp, "e%d", exponent);
+            memcpy(&r[slen+1], exp, explen+1); // Include the terminating NULL char given to us by sprintf()
+        }
+    } else {
+        memcpy(r, sbuf, last_significant_digit+1); // Include terminating NULL char.
+    }
+    if (val[0] == '-') {
+        memmove(&r[1], r, strlen(r)+1); // Be sure to include the terminating NULL char.
```

Comment:
You don't really need to do this in-place `memmove`.. Just decide what you want to move into `buf` and `memcpy` it there, since you're about to do that anyway.